modules/documentObserver.js

```diff
@@ -0,0 +1,41 @@
+'use strict';
+
+var EXPORTED_SYMBOLS = ['OnNewDocument'];
+
+var Cu = Components.utils;
+
+Cu.import('resource://gre/modules/Services.jsm');
+Cu.import('chrome://greasemonkey-modules/content/util.js');
+
+
+var callbacks = new WeakMap()
+
+function OnNewDocument(topWindow, callback) {
+	callbacks.set(topWindow, callback)
```

Comment:
It uses a weakmap, it doesn't hold strongly onto the keys (the windows in this case). So when the window goes away it can collect the callback and then it can collect the framescript in the callback's scope.